src/sinks/aws_s3/service.rs

```diff
@@ -0,0 +1,124 @@
+}
+
+impl Finalizable for S3Request {
+    fn take_finalizers(&mut self) -> EventFinalizers {
+        std::mem::take(&mut self.finalizers)
+    }
+}
+
+impl Response for PutObjectOutput {}
+
+/// Wrapper for the Rusoto S3 client.
+///
+/// Provides a `tower::Service`-compatible wrapper around the native
+/// `rusoto_s3::S3Client`, allowing it to be composed within a Tower "stack",
+/// such that we can easily and transparently provide retries, concurrency
+/// limits, rate limits, and more.
+#[derive(Clone)]
+pub struct S3Service {
+    client: S3Client,
+}
+
+impl S3Service {
+    pub const fn new(client: S3Client) -> S3Service {
+        S3Service { client }
+    }
+}
+
+impl Service<S3Request> for S3Service {
+    type Response = PutObjectOutput;
+    type Error = RusotoError<PutObjectError>;
+    type Future = BoxFuture<'static, Result<Self::Response, Self::Error>>;
+
+    fn poll_ready(&mut self, _cx: &mut Context) -> Poll<Result<(), Self::Error>> {
+        Poll::Ready(Ok(()))
+    }
+
+    fn call(&mut self, request: S3Request) -> Self::Future {
+        let options = request.options;
+
+        let content_encoding = request.content_encoding;
+        let content_encoding = options
+            .content_encoding
+            .or_else(|| content_encoding.map(|ce| ce.to_string()));
+        let content_type = options
+            .content_type
+            .or_else(|| Some("text/x-log".to_owned()));
+
+        let content_md5 = base64::encode(md5::Md5::digest(&request.body));
+
+        let mut tagging = url::form_urlencoded::Serializer::new(String::new());
+        if let Some(tags) = options.tags {
+            for (p, v) in tags {
+                tagging.append_pair(&p, &v);
+            }
+        }
+        let tagging = tagging.finish();
+
+        let body_len = request.body.len();
+        let client = self.client.clone();
+        let request = PutObjectRequest {
+            body: Some(bytes_to_bytestream(request.body)),
+            bucket: request.bucket,
+            key: request.key,
+            content_encoding,
+            content_type,
+            acl: options.acl.map(to_string),
+            grant_full_control: options.grant_full_control,
+            grant_read: options.grant_read,
+            grant_read_acp: options.grant_read_acp,
+            grant_write_acp: options.grant_write_acp,
+            server_side_encryption: options.server_side_encryption.map(to_string),
+            ssekms_key_id: options.ssekms_key_id,
+            storage_class: options.storage_class.map(to_string),
+            tagging: Some(tagging),
+            content_md5: Some(content_md5),
+            ..Default::default()
+        };
+
+        Box::pin(async move {
+            let result = client.put_object(request).in_current_span().await;
+
+            // TODO: This is fine for testing, but we should have a better
+            // pattern for this.
+            emit!(S3EventsSent {
+                byte_size: body_len,
+            });
```

Comment:
Yeah, considering it doesn't properly know how the request went it'll for sure double-emit on retries. I'm inclined to leave this as-is, comment included, and compare experiences with this PR and #8825 when Toby comes back from vacation. Then we can start the process of trying to normalize the two implementations to one another, spin off tickets for the TODOs etc.

Reply:
I'll leave this sub-conversation unresolved so it's obvious for later discussion.

Reply:
Yeah, my inclination here is that this is probably a good use case for a piece of Tower middleware that can utilize the same machinery we use to drive retries vis a vis detecting if a response is good or not, and then emitting the appropriate metrics.  This doesn't actually need to be an S3-specific metric as is, since we only know it is by the `component_type` tag anyways.

Reply:
100%